src/Stratis.SmartContracts.CLR/EmbeddedContractAddress.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System;
+using System.Linq;
+using NBitcoin;
+
+namespace Stratis.SmartContracts.CLR
+{
+    public enum EmbeddedContractType
+    {
+        Authentication = 1,
+        Multisig = 2
+    }
+
+    [AttributeUsage(AttributeTargets.Class)]
+    public class EmbeddedContractAttribute : Attribute
+    {
+        public EmbeddedContractType ContractType { get; private set; }
+
+        public EmbeddedContractAttribute(EmbeddedContractType contractType)
+        {
+            this.ContractType = contractType;
+        }
+
+        public static EmbeddedContractType GetEmbeddedContractTypeId(Type contractType)
+        {
+            return contractType.GetCustomAttributes(typeof(EmbeddedContractAttribute), true)
+                .OfType<EmbeddedContractAttribute>().FirstOrDefault().ContractType;
+        }
+    }
+
+    public struct EmbeddedContractAddress
+    {
+        private static byte[] embeddedContractSignature = new byte[] { 0, 0, 0, 0, 0, 0, 0, 0 };
+        private uint160 value;
+
+        public EmbeddedContractAddress(uint160 id)
+        {
+            this.value = id;
+        }
+
+        public EmbeddedContractAddress(Type contractType, uint version)
+        {
+            EmbeddedContractType contractTypeId = EmbeddedContractAttribute.GetEmbeddedContractTypeId(contractType);
+            this.value = new uint160(BitConverter.GetBytes(version).Concat(BitConverter.GetBytes((ulong)contractTypeId)).Concat(embeddedContractSignature).ToArray());
+        }
+
+        public uint Version { get => BitConverter.ToUInt32(this.value.ToBytes()); }
```

Comment:
Can you elaborate a little on what this means? Elsewhere in the code there seems to be a requirement for version to be equal to 1, but the two constructors for this struct seem to set `value` to very different things

Reply:
We are checking for the version being equal to 1 in the `MultiSig` and `Authentication` contracts are currently in version 1 themselves. If we do change their code it would be in a backwards compatible manner and we may still want them to exhibit their original behaviour when called by their legacy address in legacy blocks. At that point we will also be relaxing the version =1 requirement to allow the contract code to be called by its new address containing a version of 2,3,etc.